File: src/java/org/orbeon/oxf/xforms/submission/AsynchronousSubmissionManager.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xforms.submission;

import org.orbeon.oxf.common.OXFException;
import org.orbeon.oxf.util.*;
import org.orbeon.oxf.xforms.XFormsContainingDocument;
import org.orbeon.oxf.xforms.XFormsProperties;
import org.orbeon.oxf.xforms.event.XFormsEvents;
import org.orbeon.oxf.xforms.event.events.XXFormsSubmitReplaceEvent;
import org.orbeon.oxf.xforms.xbl.XBLContainer;

import java.util.Map;
import java.util.concurrent.*;

/**
 * Handle asynchronous submissions.
 *
 * The CompletionService is stored in the session, indexed by document UUID.
 *
 * See http://wiki.orbeon.com/forms/doc/developer-guide/asynchronous-submissions
 * See http://java.sun.com/j2se/1.5.0/docs/api/java/util/concurrent/ExecutorCompletionService.html
 */
public class AsynchronousSubmissionManager {

    private static final String ASYNC_SUBMISSIONS_SESSION_KEY_PREFIX = "oxf.xforms.state.async-submissions.";

    // Global thread pool
    private static final ExecutorService threadPool = Executors.newCachedThreadPool();

    private final XFormsContainingDocument containingDocument;

    public AsynchronousSubmissionManager(XFormsContainingDocument containingDocument) {
        this.containingDocument = containingDocument;
    }

    /**
     * Add a special delay event to the containing document if there are pending submissions.
     *
     * This should be called just before sending an Ajax response.
     *
     */
    public void addClientDelayEventIfNeeded() {
        if (hasPendingAsynchronousSubmissions()) {
            // NOTE: Could get isShowProgress() from submission, but default must be false
            containingDocument.addDelayedEvent(XFormsEvents.XXFORMS_POLL, containingDocument.getEffectiveId(),
                        false, false, XFormsProperties.getSubmissionPollDelay(containingDocument), true, false, null);
        }
    }

    private static String getSessionKey(XFormsContainingDocument containingDocument) {
        return getSessionKey(containingDocument.getUUID());
    }

    private static String getSessionKey(String documentUUID) {
        return ASYNC_SUBMISSIONS_SESSION_KEY_PREFIX + documentUUID;
    }

    private static AsynchronousSubmissions getAsynchronousSubmissions(boolean create, String sessionKey) {
        final Map<String, Object> sessionMap = NetUtils.getExternalContext().getSession(true).getAttributesMap();
        final AsynchronousSubmissions existingAsynchronousSubmissions = (AsynchronousSubmissions) sessionMap.get(sessionKey);
        if (existingAsynchronousSubmissions != null) {
            return existingAsynchronousSubmissions;
        } else if (create) {
            final AsynchronousSubmissions asynchronousSubmissions = new AsynchronousSubmissions();
            sessionMap.put(sessionKey, asynchronousSubmissions);
            return asynchronousSubmissions;
        } else {
            return null;
        }
    }

    public void addAsynchronousSubmission(final Callable<SubmissionResult> callable) {

        final AsynchronousSubmissions asynchronousSubmissions = getAsynchronousSubmissions(true, getSessionKey(containingDocument));

        // NOTE: If we want to re-enable foreground async submissions, we must:
        // o do a better detection: !(xf-submit-done/xf-submit-error listener) && replace="none"
        // o OR provide an explicit hint on xf:submission
        asynchronousSubmissions.submit(callable);
    }

    public boolean hasPendingAsynchronousSubmissions() {
        final AsynchronousSubmissions asynchronousSubmissions = getAsynchronousSubmissions(false, getSessionKey(containingDocument));
        return asynchronousSubmissions != null && asynchronousSubmissions.getPendingCount() > 0;
    }

    /**
     * Process all pending asynchronous submissions if any. If processing of a particular submission causes new
     * asynchronous submissions to be started, also wait for the completion of those.
     *
     * Submissions are processed in the order in which they are made available upon termination by the completion
     * service.
     *
     */
    public void processAllAsynchronousSubmissions() {
        final AsynchronousSubmissions asynchronousSubmissions = getAsynchronousSubmissions(false, getSessionKey(containingDocument));
        if (asynchronousSubmissions != null && asynchronousSubmissions.getPendingCount() > 0) {

            final IndentedLogger indentedLogger = containingDocument.getIndentedLogger(XFormsModelSubmission.LOGGING_CATEGORY);
            indentedLogger.startHandleOperation("", "processing all background asynchronous submissions");
            int processedCount = 0;
            try {
                while (asynchronousSubmissions.getPendingCount() > 0) {
                    try {
                        // Handle next completed task
                        final Future<SubmissionResult> future = asynchronousSubmissions.take();
                        final SubmissionResult result = future.get();

                        // Process response by dispatching an event to the submission
                        final XFormsModelSubmission submission = (XFormsModelSubmission) containingDocument.getObjectByEffectiveId(result.getSubmissionEffectiveId());
                        final XBLContainer container = submission.getXBLContainer(containingDocument);
                        // NOTE: not clear whether we should use an event for this as there doesn't seem to be a benefit
                        container.dispatchEvent(new XXFormsSubmitReplaceEvent(containingDocument, submission, result));

                    } catch (Throwable throwable) {
                        // Something bad happened
                        throw new OXFException(throwable);
                    }

                    processedCount++;
                }
            } finally {
                indentedLogger.endHandleOperation("processed", Integer.toString(processedCount));
            }
        }
    }

    /**
     * Process all completed asynchronous submissions if any. This method returns as soon as no completed submission is
     * available.
     *
     * Submissions are processed in the order in which they are made available upon termination by the completion
     * service.
     *
     */
    public void processCompletedAsynchronousSubmissions() {
        final AsynchronousSubmissions asynchronousSubmissions = getAsynchronousSubmissions(false, getSessionKey(containingDocument));
        if (asynchronousSubmissions != null && asynchronousSubmissions.getPendingCount() > 0) {
            final IndentedLogger indentedLogger = containingDocument.getIndentedLogger(XFormsModelSubmission.LOGGING_CATEGORY);
            indentedLogger.startHandleOperation("", "processing completed background asynchronous submissions");

            int processedCount = 0;
            try {
                Future<SubmissionResult> future = asynchronousSubmissions.poll();
                while (future != null) {
                    try {
                        // Handle next completed task
                        final SubmissionResult result = future.get();

                        // Process response by dispatching an event to the submission
                        final XFormsModelSubmission submission = (XFormsModelSubmission) containingDocument.getObjectByEffectiveId(result.getSubmissionEffectiveId());
                        final XBLContainer container = submission.getXBLContainer(containingDocument);
                        // NOTE: not clear whether we should use an event for this as there doesn't seem to be a benefit
                        container.dispatchEvent(new XXFormsSubmitReplaceEvent(containingDocument, submission, result));
                    } catch (Throwable throwable) {
                        // Something bad happened
                        throw new OXFException(throwable);
                    }

                    processedCount++;

                    future = asynchronousSubmissions.poll();
                }
            } finally {
                indentedLogger.endHandleOperation("processed", Integer.toString(processedCount),
                        "pending", Integer.toString(asynchronousSubmissions.getPendingCount()));
            }
        }
    }

    private static class AsynchronousSubmissions {
        private final CompletionService<SubmissionResult> completionService = new ExecutorCompletionService<SubmissionResult>(threadPool);
        private int pendingCount = 0;

        public Future<SubmissionResult> submit(Callable<SubmissionResult> task) {
            final Future<SubmissionResult> future = completionService.submit(task);
            pendingCount++;
            return future;
        }

        public Future<SubmissionResult> poll() {
            final Future<SubmissionResult> future = completionService.poll();
            if (future != null)
                pendingCount--;
            return future;
        }

        public Future<SubmissionResult> take() throws InterruptedException {
            final Future<SubmissionResult> future = completionService.take();
            pendingCount--;
            return future;
        }

        public int getPendingCount() {
            return pendingCount;
        }
    }
}

File: xforms/jvm/src/main/scala/org/orbeon/oxf/xforms/submission/AsynchronousSubmissionManager.scala
/**
  * Copyright (C) 2010 Orbeon, Inc.
  *
  * This program is free software; you can redistribute it and/or modify it under the terms of the
  * GNU Lesser General Public License as published by the Free Software Foundation; either version
  * 2.1 of the License, or (at your option) any later version.
  *
  * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
  * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  * See the GNU Lesser General Public License for more details.
  *
  * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
  */
package org.orbeon.oxf.xforms.submission

import java.io.{Externalizable, ObjectInput, ObjectOutput}
import java.util.concurrent._
import javax.enterprise.concurrent.ManagedExecutorService
import javax.naming.{InitialContext, NamingException}

import org.orbeon.oxf.externalcontext.{AsyncRequest, LocalExternalContext}
import org.orbeon.oxf.pipeline.InitUtils._
import org.orbeon.oxf.pipeline.api.PipelineContext
import org.orbeon.oxf.util.Logging._
import org.orbeon.oxf.util.{IndentedLogger, NetUtils}
import org.orbeon.oxf.xforms.XFormsContainingDocument
import org.orbeon.oxf.xforms.event.XFormsEvents

/**
  * Handle asynchronous submissions.
  *
  * The `CompletionService` is stored in the session, indexed by document UUID.
  *
  * See https://doc.orbeon.com/xforms/submission-asynchronous.html
  * See http://java.sun.com/j2se/1.5.0/docs/api/java/util/concurrent/ExecutorCompletionService.html
  */
class AsynchronousSubmissionManager(val containingDocument: XFormsContainingDocument) {

  import AsynchronousSubmissionManager._

  private implicit def indentedLogger: IndentedLogger =
    containingDocument.getIndentedLogger(XFormsModelSubmission.LOGGING_CATEGORY)

  /**
    * Add a special delay event to the containing document if there are pending submissions.
    *
    * This should be called just before sending an Ajax response.
    */
  def addClientDelayEventIfNeeded(): Unit =
    if (hasPendingAsynchronousSubmissions)
      containingDocument.addDelayedEvent(
        eventName         = XFormsEvents.XXFORMS_POLL,
        targetEffectiveId = containingDocument.getEffectiveId,
        bubbles           = false,
        cancelable        = false,
        time              = System.currentTimeMillis + containingDocument.getSubmissionPollDelay,
        discardable       = true,
        showProgress      = false, // could get from submission, but default must be `false`
        allowDuplicates   = false  // no need for duplicates
      )

  def addAsynchronousSubmission(callable: Callable[SubmissionResult]): Unit = {
    val asynchronousSubmissionsOpt =
      findAsynchronousSubmissions(
        create = true,
        sessionKey(containingDocument)
      )

    // NOTE: If we want to re-enable foreground async submissions, we must:
    // - do a better detection: !(xf-submit-done/xf-submit-error listener) && replace="none"
    // - OR provide an explicit hint on xf:submission
    asynchronousSubmissionsOpt foreach {
      _.submit(
        {
          // Make sure this is created at the time `submit` is called
          // Should we use `AsyncExternalContext` here?
          val newExternalContext = {

              val currentExternalContext = NetUtils.getExternalContext

              new LocalExternalContext(
                currentExternalContext.getWebAppContext,
                new AsyncRequest(currentExternalContext.getRequest),
                currentExternalContext.getResponse
              )
            }

          () ⇒ withPipelineContext { pipelineContext ⇒
            pipelineContext.setAttribute(PipelineContext.EXTERNAL_CONTEXT, newExternalContext)
            callable.call()
          }
        }
      )
    }
  }

  def hasPendingAsynchronousSubmissions: Boolean = {

    val asynchronousSubmissionsOpt =
      findAsynchronousSubmissions(
        create = false,
        sessionKey(containingDocument)
      )

    asynchronousSubmissionsOpt exists (_.pendingCount > 0)
  }

  /**
    * Process all pending asynchronous submissions if any. If processing of a particular submission causes new
    * asynchronous submissions to be started, also wait for the completion of those.
    *
    * Submissions are processed in the order in which they are made available upon termination by the completion
    * service.
    */
  def processAllAsynchronousSubmissions(): Unit = {

    val asynchronousSubmissionOpt =
      findAsynchronousSubmissions(
        create = false,
        sessionKey(containingDocument)
      )

    asynchronousSubmissionOpt filter (_.pendingCount > 0) foreach { asynchronousSubmission ⇒

      withDebug("processing all background asynchronous submissions") {
        var processedCount = 0
        try {
          while (asynchronousSubmission.pendingCount > 0) {

            // Handle next completed task
            val result = asynchronousSubmission.take().get()

            // Process response by dispatching an event to the submission
            val submission =
              containingDocument.getObjectByEffectiveId(result.getSubmissionEffectiveId).asInstanceOf[XFormsModelSubmission]

            submission.doSubmitReplace(result)

            processedCount += 1
          }
        } finally
          debugResults(List("processed" → processedCount.toString))
      }
    }
  }

  /**
    * Process all completed asynchronous submissions if any. This method returns as soon as no completed submission is
    * available.
    *
    * Submissions are processed in the order in which they are made available upon termination by the completion
    * service.
    */
  def processCompletedAsynchronousSubmissions(): Unit = {

    val asynchronousSubmissionsOpt =
      findAsynchronousSubmissions(
        create = false,
        sessionKey(containingDocument)
      )

    asynchronousSubmissionsOpt filter (_.pendingCount > 0) foreach { asynchronousSubmissions ⇒
      withDebug("processing completed background asynchronous submissions") {
        var processedCount = 0
        try {
          var future = asynchronousSubmissions.poll()
          while (future.isDefined) {

            val result = future.get.get()

            val submission =
              containingDocument.getObjectByEffectiveId(result.getSubmissionEffectiveId).asInstanceOf[XFormsModelSubmission]

            submission.doSubmitReplace(result)

            processedCount += 1

            future = asynchronousSubmissions.poll()
          }
        } finally
          debugResults(
            List(
              "processed" → processedCount.toString,
              "pending"  → asynchronousSubmissions.pendingCount.toString
            )
          )
      }
    }
  }
}

private object AsynchronousSubmissionManager {

  val AsyncSubmissionsSessionKeyPrefix = "oxf.xforms.state.async-submissions."

  // Global thread pool if none provided by the app server
  private lazy val threadPool = Executors.newCachedThreadPool

  private def getExecutorService: ExecutorService =
    try {
      // If the app server gives us an `ExecutorService` (e.g. with WildFly), use it
      // (See §EE.5.21, page 146 of the Java EE 7 spec)
      InitialContext.doLookup[ManagedExecutorService]("java:comp/DefaultManagedExecutorService")
    } catch {
      case _: NamingException ⇒
        // If no `ExecutorService` is provided by the app server (e.g. with Tomcat), use our global thread pool
        threadPool
    }

  def sessionKey(doc: XFormsContainingDocument): String = AsyncSubmissionsSessionKeyPrefix + doc.getUUID

  def findAsynchronousSubmissions(create: Boolean, sessionKey: String): Option[AsynchronousSubmissions] = {

    val session = NetUtils.getExternalContext.getRequest.getSession(true)

    session.getAttribute(sessionKey) map (_.asInstanceOf[AsynchronousSubmissions]) orElse {
      if (create) {
        val asynchronousSubmissions = new AsynchronousSubmissions
        session.setAttribute(sessionKey, asynchronousSubmissions)
        Some(asynchronousSubmissions)
      } else
        None
    }
  }

  class AsynchronousSubmissions extends Externalizable {

    private val completionService = new ExecutorCompletionService[SubmissionResult](getExecutorService)

    private var _pendingCount = 0
    def pendingCount = _pendingCount

    def submit(task: () ⇒ SubmissionResult): Future[SubmissionResult] = {

      val future = completionService.submit(
        new Callable[SubmissionResult]() {
          def call() = task()
        }
      )

      _pendingCount += 1
      future
    }

    def poll(): Option[Future[SubmissionResult]] =
      Option(completionService.poll()) map { f ⇒
        _pendingCount -= 1
        f
      }

    def take(): Future[SubmissionResult] = {
      val f = completionService.take()
      _pendingCount -= 1
      f
    }

    // So that this can be stored in session that require `Serializable`
    def writeExternal(out: ObjectOutput) = ()
    def readExternal(in: ObjectInput)    = ()
  }

}

